userspace/libsinsp/utils.cpp

```diff
@@ -595,172 +596,31 @@ bool sinsp_utils::sockinfo_to_str(sinsp_sockinfo* sinfo, scap_fd_type stype, cha
+    auto p2 = std::filesystem::path(path2, std::filesystem::path::format::generic_format);
 
-//
-// Args:
-//  - target: the string where we are supposed to start copying
-//  - targetbase: the base of the path, i.e. the furthest we can go back when
-//                following parent directories
-//  - path: the path to copy
-//
-void copy_and_sanitize_path(char* target, char* targetbase, const char* path, char separator)
-{
-	char* tc = target;
-	const char* pc = path;
-	g_invalidchar ic;
+	auto path_concat = (p1 / p2).lexically_normal();
+	std::string result = path_concat;
 
-	while(true)
+	//
+	// If the path ends with a separator, remove it, as the OS does.
+	//
+	if (result.length() > 1 && result.back() == '/')
 	{
-		if(*pc == 0)
-		{
-			*tc = 0;
-
-			//
-			// If the path ends with a separator, remove it, as the OS does.
-			//
-			if((tc > (targetbase + 1)) && (*(tc - 1) == separator))
-			{
-				*(tc - 1) = 0;
-			}
-
-			return;
-		}
-
-		if(ic(*pc))
-		{
-			//
-			// Invalid char, substitute with a '.'
-			//
-			*tc = '.';
-			tc++;
-			pc++;
-		}
-		else
-		{
-			//
-			// If path begins with '.' or '.' is the first char after a '/'
-			//
-			if(*pc == '.' && (tc == targetbase || *(tc - 1) == separator))
-			{
-				//
-				// '../', rewind to the previous separator
-				//
-				if(*(pc + 1) == '.' && *(pc + 2) == separator)
-				{
-					rewind_to_parent_path(targetbase, &tc, &pc, 3);
-				}
-				//
-				// '..', with no separator.
-				// This is valid if we are at the end of the string, and in that case we rewind.
-				//
-				else if(*(pc + 1) == '.' && *(pc + 2) == 0)
-				{
-					rewind_to_parent_path(targetbase, &tc, &pc, 2);
-				}
-				//
-				// './', just skip it
-				//
-				else if(*(pc + 1) == separator)
-				{
-					pc += 2;
-				}
-				//
-				// '.', with no separator.
-				// This is valid if we are at the end of the string, and in that case we rewind.
-				//
-				else if(*(pc + 1) == 0)
-				{
-					pc++;
-				}
-				//
-				// Otherwise, we leave the string intact.
-				//
-				else
-				{
-					*tc = *pc;
-					pc++;
-					tc++;
-				}
-			}
-			else if(*pc == separator)
-			{
-				//
-				// separator, if the last char is already a separator, skip it
-				//
-				if(tc > targetbase && *(tc - 1) == separator)
-				{
-					pc++;
-				}
-				else
-				{
-					*tc = *pc;
-					tc++;
-					pc++;
-				}
-			}
-			else
-			{
-				//
-				// Normal char, copy it
-				//
-				*tc = *pc;
-				tc++;
-				pc++;
-			}
-		}
+		result.pop_back();
 	}
-}
 
-//
-// Return false if path2 is an absolute path
-//
-bool sinsp_utils::concatenate_paths(char* target,
-									uint32_t targetlen,
-									const char* path1,
-									uint32_t len1,
-									const char* path2,
-									uint32_t len2)
-{
-	if(targetlen < (len1 + len2 + 1))
+	if (max_len != -1 && result.length() > max_len)
 	{
-		strlcpy(target, "/PATH_TOO_LONG", targetlen);
-		return false;
+		return "/PATH_TOO_LONG";
```

Comment:
Given that we just return a bogus path with no error condition (exception or error code), does it really make sense to enforce a max path? Perhaps we can remove the concept altogether. The `std::filesystem` library doesn't have it, for instance.

Reply:
Besides libs there is also a safety guard aspect wrt not outputting too large strings in the event logs. But I am also not sure of returning a bogus string versus just resizing it.

Reply:
I did think about this a little. I don't like the fact that we return this kind of string to signal a potential error but in the code there currently is no guarantee that the buffers that these strings will be copied into (incl. later during processing) are never plain old C-style `char buf[SCAP_MAX_PATH_SIZE]` (I didn't check any of those instances but it's possible that we still have some). In that case, keeping the input string will lead to truncation (with e.g. an `snprintf` or `strlcpy`).
Truncating a path string can potentially lead to security issues so this is not a behavior that I would change right now.

However, from what you're seeing in this PR and similar ones we are trying to replace those plain old `char buf`s with `std::string`, meaning that we're getting closer to a time in which we can actually check all the locations and be happy about the fact that we're using proper strings, thus being able to remove this limitation. You probably also noticed that the parameter is optional, so we will be able to relatively painlessly remove that extra argument without having to change all callers.

Either way, I added a comment to call out this behavior in the function description.